warehouse/legacy/api/pypi.py

```diff
@@ -0,0 +1,478 @@
+                    "summary", "description", "license",
+                    "author", "author_email", "maintainer", "maintainer_email",
+                    "keywords", "platform",
+                    "home_page", "download_url",
+                    "requires_python",
+                }
+            }
+        )
+        request.db.add(release)
+
+    # TODO: We need a better solution to this than to just do it inline inside
+    #       this method. Ideally the version field would just be sortable, but
+    #       at least this should be some sort of hook or trigger.
+    releases = (
+        request.db.query(Release)
+                  .filter(Release.project == project)
+                  .all()
+    )
+    for i, r in enumerate(sorted(
+            releases, key=lambda x: packaging.version.parse(x.version))):
+        r._pypi_ordering = i
+
+    # Pull the filename out of our POST data.
+    filename = request.POST["content"].filename
+
+    # Make sure that the filename does not contain and path seperators.
+    if "/" in filename or "\\" in filename:
+        raise _exc_with_message(
+            HTTPBadRequest,
+            "Cannot upload a file with '/' or '\\' in the name.",
+        )
+
+    # Make sure the filename ends with an allowed extension.
+    if _dist_file_re.search(filename) is None:
+        raise _exc_with_message(HTTPBadRequest, "Invalid file extension.")
+
+    # Make sure that our filename matches the project that it is being uploaded
+    # to.
+    prefix = pkg_resources.safe_name(project.name).lower()
+    if not pkg_resources.safe_name(filename).lower().startswith(prefix):
+        raise _exc_with_message(
+            HTTPBadRequest,
+            "The filename for {!r} must start with {!r}.".format(
+                project.name,
+                prefix,
+            )
+        )
+
+    # Check to see if the file that was uploaded exists in our filename log.
+    if (request.db.query(
+            request.db.query(Filename)
+                      .filter(Filename.filename == filename)
+                      .exists()).scalar()):
+        raise _exc_with_message(
+            HTTPBadRequest,
+            "This filename has previously been used, you should use a "
+            "different version.",
+        )
+
+    # Buffer the entire file into memory, checking the hash of the file as we
+    # go along.
+    file_content = b""
+    file_size = 0
+    file_hash = hashlib.md5()
+    for chunk in iter(lambda: request.POST["content"].file.read(8096), b""):
+        file_size += len(chunk)
+        if file_size > MAXIMUM_FILESIZE:
+            raise _exc_with_message(HTTPBadRequest, "File too large.")
+        file_content += chunk
+        file_hash.update(chunk)
+
+    # Get the signature if it was included.
+    if "gpg_signature" in request.POST:
+        signature = b""
+        for chunk in iter(
+                lambda: request.POST["gpg_signature"].read(8096), b""):
+            signature += chunk
```

Comment:
Same (you can just read the whole thing here, you're not doing anything with each chunk)